File: didcomm-demo-python/didcomm_demo/didcomm_cli.py
import click
from peerdid.types import DIDDocVerMaterialFormat

from didcomm_demo.didcomm_demo import DIDCommDemo
from didcomm_demo.secrets.secrets_resolver_demo import SecretsResolverDemo

secrets_resolver = SecretsResolverDemo()


def set_secrets_resolver(resolver: SecretsResolverDemo):
    global secrets_resolver
    secrets_resolver = resolver


@click.group()
def cli():
    pass


@cli.command()
@click.option('--auth_keys_count', default=1, help='Number of authentication keys')
@click.option('--agreement_keys_count', default=0, help='Number of agreement keys')
@click.option('--service_endpoint', default=None, help='Optional service endpoint')
@click.option('--service_routing_keys', default=[], help='Optional service routing keys')
def create_peer_did(auth_keys_count, agreement_keys_count, service_endpoint, service_routing_keys):
    did = DIDCommDemo(secrets_resolver).create_peer_did(
        auth_keys_count=auth_keys_count,
        agreement_keys_count=agreement_keys_count,
        service_endpoint=service_endpoint,
        service_routing_keys=service_routing_keys
    )
    click.echo(f"{did}")


@cli.command()
@click.argument('did')
@click.option('--format', type=click.Choice(['jwk', 'multibase'], case_sensitive=False),
              default="jwk",
              help='DID Doc format (JWK or Multibase)')
def resolve_peer_did(did, format):
    format = DIDDocVerMaterialFormat.JWK if format == "jwk" else DIDDocVerMaterialFormat.MULTIBASE
    did_doc_json = DIDCommDemo.resolve_peer_did(did, format)
    click.echo(f"{did_doc_json}")


@cli.command()
def pack():
    pass


@cli.command()
def unpack():
    pass

# if __name__ == '__main__':
#     #cli()
#     private_key_jwk_dict, public_key_jwk_dict = generate_ed25519_key()
#     inception_key = jwk_x25519_to_peer_did_auth_key(public_key_jwk_dict)
#     did = peer_did.create_peer_did_numalgo_0(inception_key)

File: didcomm-demo-jvm/app/src/main/kotlin/org/dif/didcomm/demo/DIDCommCLI.kt
@file:OptIn(ExperimentalCli::class)

package org.dif.didcomm.demo


import kotlinx.cli.*
import org.dif.exceptions.DIDCommException
import org.dif.peerdid.core.DIDDocVerMaterialFormat


val demo = DIDCommDemo()

class PeerDIDCreatorCommand : Subcommand("create-peer-did", "Creates a new Peer DID and corresponding secrets") {

    val authKeysCount by option(
        ArgType.Int,
        description = "Number of authentication keys",
        fullName = "auth-keys-count"
    ).default(1)
    val agreementKeysCount by option(
        ArgType.Int,
        description = "Number of agreement keys",
        fullName = "agreement-keys-count"
    ).default(1)


    override fun execute() {
        val res = try {
            demo.createPeerDID(
                authKeysCount = authKeysCount, agreementKeysCount = agreementKeysCount
            )
        } catch (e: IllegalArgumentException) {
            e.localizedMessage
        }
        println()
        println(res)
        println()
    }

}

enum class Format {
    JWK,
    BASE58,
    MULTIBASE
}

class ResolvePeerDIDCommand : Subcommand("resolve-peer-did", "Resolve a Peer DID to DID Doc JSON") {

    val did by argument(ArgType.String, description = "Peer DID to be resolved")
    val format by option(
        ArgType.Choice<Format>(),
        shortName = "f",
        description = "Peer DID to be resolved"
    ).default(Format.JWK)

    override fun execute() {
        val res = try {
            DIDCommDemo.resolvePeerDID(
                did,
                format = when (format) {
                    Format.JWK -> DIDDocVerMaterialFormat.JWK
                    Format.BASE58 -> DIDDocVerMaterialFormat.BASE58
                    Format.MULTIBASE -> DIDDocVerMaterialFormat.MULTIBASE
                }
            )
        } catch (e: Exception) {
            e.localizedMessage
        }
        println()
        println(res)
        println()
    }

}

class PackCommand : Subcommand("pack", "Packs the message") {

    val message by argument(ArgType.String, description = "Message to pack")
    val to by option(ArgType.String, description = "Receiver's DID").required()
    val from by option(ArgType.String, description = "Sender's DID. Anonymous encryption is used if not set.")
    val signFrom by option(
        ArgType.String,
        fullName = "sign-from",
        description = "Sender's DID for optional signing. The message is not signed if not set."
    )
    val protectSender by option(
        ArgType.Boolean,
        fullName = "protect-sender",
        description = "Whether the sender's ID (DID) must be hidden. True by default."
    ).default(true)

    override fun execute() {
        val res = try {
            demo.pack(
                data = message, to = to, from = from, signFrom = signFrom, protectSender = protectSender
            ).packedMessage
        } catch (e: DIDCommException) {
            e.localizedMessage
        }
        println()
        println(res)
        println()
    }

}

class UnpackCommand : Subcommand("unpack", "Unpacks the message") {

    val message by argument(ArgType.String, description = "Message to unpack")

    override fun execute() {
        val res = try {
            val unpackRes = demo.unpack(message)
            unpackRes.from?.let {
                "authcrypted ${unpackRes.message} from ${unpackRes.from} to ${unpackRes.to}"
            } ?: {
                "anoncrypted ${unpackRes.message} to ${unpackRes.to}"
            }
        } catch (e: DIDCommException) {
            e.localizedMessage
        }
        println()
        println(res)
        println()
    }

}

fun main(args: Array<String>) {
    val parser = ArgParser("didcomm-cli")
    parser.subcommands(PeerDIDCreatorCommand(), ResolvePeerDIDCommand(), PackCommand(), UnpackCommand())
    parser.parse(args)
}
